core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -126,13 +149,23 @@ public int read(byte[] b, int off, int len) throws IOException {
 
     int bytesLeft = len;
     int currentOffset = off;
+    CountingRetry retry = new CountingRetry(MAX_WORKERS_TO_RETRY);
     while (bytesLeft > 0 && mPosition != mLength) {
       updateStream();
-      int bytesRead = mBlockInStream.read(b, currentOffset, bytesLeft);
-      if (bytesRead > 0) {
-        bytesLeft -= bytesRead;
-        currentOffset += bytesRead;
-        mPosition += bytesRead;
+      try {
```

Comment:
ditto. Same here, it is quite hard to reason which is the retry body. Can we have similar pattern as my above comment? I know the `reset` may complicate the story. But it is quite hard for me to tell indeed which part we retry.

Reply:
I am debating on this one. I want to follow the while(retry.attemptRetry()) pattern, but that means we will need to complicate existing loop condition:
```
while (bytesLeft > 0 && mPosition != mLength && shouldRetry && retry.attemptRetry()) {
  ...
  retry.reset();
  shouldRetry = true;
  ...
}
```

or add a new nested loop:
```
while (bytesLeft > 0 && mPosition != mLength) {
  while(shouldRetry && retry.attemptRetry()) {
    ...
    retry.reset();
    shouldRetry = true;
    ...
  }
}
```
I chose the current solution because it feels like either combining the while condition or nested while loop makes it less readable. It is also nice to have the retry logic close to where it handles the exception to retry. Let me know if you have a strong preference on one pattern over the others.

Reply:
I actually don't mind the pattern of using
```java
while (true) {
  try {
    // do something
  } catch (something) {
    if (!retry.shouldRetry()) {
      // clean up and exit
    }
    // recover and retry
  }
}
```
It's nice that it gives a place to have code that only triggers when we run out of retries. Otherwise it's a little awkward after the `while (retry.shouldRetry()` loop to go back and figure out whether we exited due to retries or success. We also probably need to let an exception escape the `while` scope so that we can rethrow it.

FYI there is currently an effort to improve the usability of the retry mechanism: https://github.com/Alluxio/alluxio/pull/6895. Once we've settled on the cleanest pattern, we can apply it here, probably in a later PR or as a part of 6895

Reply:
SG @aaudiber . I saw that PR too. let's move the discussion there